runtime/compiler/trj9/env/VMJ9.cpp

```diff
@@ -1292,7 +1298,11 @@ uintptrj_t
 TR_J9VMBase::getReferenceFieldAt(uintptrj_t objectPointer, uintptrj_t fieldOffset)
    {
    TR_ASSERT(haveAccess(), "Must haveAccess in getReferenceFieldAt");
-   return getReferenceFieldAtAddress(objectPointer + sizeof(J9Object) + fieldOffset);
+   if (TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads())
```

Comment:
After the change getReferenceFieldAt and getVolatileReferenceFieldAt will be effectively equivalent, and one should go away.

Reply:
Right, my bad. I was modifying `TR_J9VMBase::getReferenceElement` then I was side tracked to `getReferenceFieldAtAddress`. Working on the change

Reply:
Many methods in this class duplicates the knowledge of object layout (compressed ref or not, contiguous arrays or not etc) and dynamics (barrier needed or not) , that is effectively part of core VM (of course, the class name is TR_J9VMBase). For many of these things there is already API or macro that exists elsewhere in core VM (like j9accessbarrier.h) that are aware of all these variables. Unless there is a performance aspect to optimize specific case, or if API happen not to exist, you should just pass the call to those core VM APIs/macros.

For original purpose of this change (adding awareness of read barrier) you need to treat getReferenceFieldAt (just pass it to  J9OBJECT_OBJECT_LOAD) and getReferenceElement  (just pass it to J9JAVAARRAYOFOBJECT_LOAD).

But there are others that outside of scope of this change, that would be wise to simplify, like getInt32Element (just pass it to J9JAVAARRAYOFINT_LOAD), the one about array length, but some others too.

One case that of non existent macro is J9OBJECT__PRE_OBJECT_LOAD, but it should really exist., and then used in getReferenceFieldAtAddress. Instead of waiting for one to be created,  its ok to just call J9ReadBarrier (as you already did).

Interestingly, there seems to be nothing in this class that does a plain 'store' (there are some compare-and-swaps, though), so there is no need for any extra write barrier awareness (no surprise, otherwise, our concurrent marking GCs would already have experienced problems).

And hopefully the last thing, perhaps it's ok to keep both  getReferenceFieldAt,  getVolatileReferenceFieldAt, they do differ in the value of the last argument (hard coded false vs IS_VOLATILE()).

Reply:
> For many of these things there is already API or macro that exists elsewhere in core VM (like j9accessbarrier.h) that are aware of all these variables. Unless there is a performance aspect to optimize specific case, or if API happens not to exist, you should just pass the call to those core VM APIs/macros.

I definitely agree! Most JIT developers have not stepped out of the JIT boundary so we are mostly ignorant when it comes to all the APIs the VM supports. I saw a bunch of stuff related to strings that we are re-inventing in the JIT instead of calling well defined VM APIs. We need to do a better job moving forward. Reviews like this are a great resource for us to learn about existing APIs so thanks for this!